### ISSUE TITLE
Clear the data/log for jobs in Gitea Actions that are too old

### ISSUE THREAD

**Issue by wolfogre** [go-gitea/gitea#24256]
### Feature Description

Jobs that have been done for too long no longer need to retain data/log, so it is necessary to regularly clean them up.

- Log file: #31735
- Outputs.
- Artifacts: #26131
- Old raws in job/run/task tables
- And more?

### Details

https://github.com/go-gitea/gitea/pull/26275#issuecomment-1667072215 : Some logic which could help:

- What you want delete is a `run`, so you need to delete a row of `ActionRun`.
- A `run` could include multiple `jobs`, so you need to delete multiple rows of `ActionRunJob`.
- When a `job` has been picked up by a runner to execute, there will be a `task`, however, a `job` could be rerun many times, so you need to delete multiple rows of `ActionTask`.
- Every `task` could include
  - multiple `steps`, so `ActionTaskStep`
  - multiple `outputs`, so `ActionTaskOutput`
- The log of the `task` is stored as a file (in DBFS or storage), so you need to detele it according to `ActionTask.LogFilename` or `ActionTask.LogInStorage`.
- In addition, there may be some `artifacts` uploaded to the `run`, so you need to delete multiple rows of `ActionArtifact`.
  - The files of `artifacts` are stored in storage, so you need to remove them.
- Finally, update `NumActionRuns` and `NumClosedActionRuns` of `Repository`.

**Comment by yp05327** [comment]
Maybe a related question?
To save the storage usages, is it worth compressing these data/log?

**Comment by ghnp5** [comment]
I was now looking for this 😊
Thanks!!

**Comment by lunny** [comment]
> 

Maybe we can compress as zip, so that when display via HTTP, it can be unzip by web browser.

**Comment by ghnp5** [comment]
Maybe, but I think most important is to be able to purge too, with some Retention setting (e.g. older than 1 month, keep last 3, etc), so that the storage doesn't grow forever.

**Comment by nicorac** [comment]
+1 for purge ability; this is my use case:

I'm managing an interlal Gitea server, and I'm going to make Actions available to my co-worker devs.

Since there's no other way to test push Actions than... pushing, I've played with them by "trial and error" 😉.
These tests caused the repo to actually contain 150+ **completely useless** job executions, so I'd like to purge all of them before going to "production".

To make it worst, I suppose that once Actions will be made available to anyone on my server, my co-workers will also start testing them with the same result.

And finally, once in production, there's no need for us to keep any of repos Actions logs longer then 2wks...

**Comment by diamante0018** [comment]
about the Artifacts would it be possible to add a little rubbish icon next to the file name so they can be deleted manually? (talking about artifacts created by actions and uploaded via actions/upload-artifact)

**Comment by bilogic** [comment]
For me, 7 days is more than enough

**Comment by f1mishutka** [comment]
> For me, 7 days is more than enough

Would be much better to have this interval configurable per repository.

**Comment by bilogic** [comment]
![image](https://github.com/go-gitea/gitea/assets/946010/45b36213-766b-4d77-8ffa-d1d52f049652)

**Comment by bilogic** [comment]
![image](https://github.com/go-gitea/gitea/assets/946010/67088bda-37dd-43aa-9536-1d9d4e611aea)

After deleting the old action logs, the spinners don't stop and there are no messages to indicate the log no longer exists.

**Comment by diamante0018** [comment]
> ![image](https://private-user-images.githubusercontent.com/946010/316351574-67088bda-37dd-43aa-9536-1d9d4e611aea.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEzMjQwMTcsIm5iZiI6MTcxMTMyMzcxNywicGF0aCI6Ii85NDYwMTAvMzE2MzUxNTc0LTY3MDg4YmRhLTM3ZGQtNDNhYS05NTM2LTFkOWQ0ZTYxMWFlYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyNFQyMzQxNTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hNGIwYzMzNTI3NTJjOGE0MWZiMmUyMGM1NjZiZGU5ZTQzMjc4ZDZiMTBlNDEwODMyMTQwNDlmNDBiMjUzYjlhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Q-su00WUrmn54f9MdWrpQDQUJtrOsPP_S66tvRivsqU)
> 
> After deleting the old action logs, the spinners don't stop and there are no messages to indicate the log no longer exists.

Append to me as well when the runner dies because it could not download an action.
So the wheels keep spinning forever

**Comment by wolfogre** [comment]
> After deleting the old action logs, the spinners don't stop and there are no messages to indicate the log no longer exists.

@bilogic How do you clear the old action logs? Just remove them from disk/storage? If so, I would say it's safe to do that to free up space. However, the old jobs cannot be displayed properly on the UI since there's no code for the inconsistency.

**Comment by bilogic** [comment]
@wolfogre Yes, I just removed from storage, but that was because I had a backup. And yes, my image was to illustrate the need to report an error if the logs are deleted.

**Comment by bluejack2000** [comment]
Any updates on this? Our action view grows and grows. We have over 20k action runs at the moment and it is very laggy in the UI (page control shows 706 pages of action runs). Also the action_task table is about 600MB+. We need that cleaned up.

**Comment by lunny** [comment]
#31735 introduced a method to delete the old logs automatically.

**Comment by ghnp5** [comment]
@lunny - that deletes the logs, but not the actual Action runs.

**Comment by gpxinrt** [comment]
This would also be a useful feature for cleaning up the stuck runs from #35780  because they can't be deleted even after implementing the 'fix'.

**Comment by ptrdlbrg** [comment]
There is a workaround for this issue now that individual workflow runs can be deleted via the API. I'd like to see this implemented natively in Gitea, but until then I'll use my own workaround. I provide this workaround for the benefit of others who would like to bulk delete old workflow runs. 

The bash code below will delete workflow runs for a repo that are older than the specified number of days. The API token needs to have read/write privileges for the repo. The commands `curl` and `jq` need to be installed.

```
RETENTION_DAYS=60
GITEA_HOST=gitea.example.com
GITEA_TOKEN=your_token_several_hex_digits_aabbccddee
REPO_OWNER=repo-owner
REPO_NAME=your-repo-name
curl -s --request GET --header "authorization: token ${GITEA_TOKEN}" --url "https://${GITEA_HOST}/api/v1/repos/${REPO_OWNER}/${REPO_NAME}/actions/runs" \
  | jq --arg the_date $(date -d "$RETENTION_DAYS days ago" +'%Y-%m-%d') '.workflow_runs[] | select((.started_at < $the_date and (.conclusion=="success" or .conclusion=="failure")) or (.completed_at < $the_date and .conclusion=="cancelled")) | .id' \
  | xargs -i'{}' --no-run-if-empty -n 1 \
      curl -s -X DELETE -H "authorization: token ${GITEA_TOKEN}" -w "Run ID: {}, Status: %{http_code}\n" -o /dev/null --url "https://${GITEA_HOST}/api/v1/repos/${REPO_OWNER}/${REPO_NAME}/actions/runs/{}"
```

Explanation:
- Set the environment variables for your Gitea host, API token, Repo owner, and repo name. Specify how many days old a workflow run is retained for.
- Make a `curl` request to output the workflow runs for the repo.
- Pipe the JSON output from the `curl` command to  `jq` to filter the list of workflow runs. A `date` command is used to output a date that is `RETENTION_DAYS` days before today. That date is used as an argument to the `jq` command. The `jq` filter factors in the run `conclusion` because the the `started_at` date is only valid for success or failure runs. Cancelled runs only have a valid `completed_at` date.  The `jq` command outputs the run id for each run that is eligible for deletion.
- Pipe IDs from the `jq` output into an `xargs` command. The `xargs` command will run a `curl` command for each ID from the input. `xargs` will replace any reference to `{}` with the ID obtained from the input.
- The `curl` command executes a DELETE for the workflow run with the supplied ID. The command writes any output to `/dev/null` so the request body is not sent to stdout. Each `curl` run outputs which Run ID it is operating on and the HTTP status code. 

USE THIS CODE AT YOUR OWN RISK. It works for me, but please test the output of the `jq` command before piping the data to the last `xargs`/`curl` command that actually performs deletes.

**Comment by ghnp5** [comment]
Wondering if there is any news about this one? 😊
Many thanks!
